File: qa_portfolio/pages/inventory_page.py
"""Inventory (products) page POM for saucedemo.com."""

from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import Select
from .base_page import BasePage


class InventoryPage(BasePage):
    INVENTORY_CONTAINER = (By.ID, "inventory_container")
    INVENTORY_ITEMS     = (By.CLASS_NAME, "inventory_item")
    ITEM_NAMES          = (By.CLASS_NAME, "inventory_item_name")
    ITEM_PRICES         = (By.CLASS_NAME, "inventory_item_price")
    SORT_DROPDOWN       = (By.CLASS_NAME, "product_sort_container")
    CART_BADGE          = (By.CLASS_NAME, "shopping_cart_badge")
    CART_ICON           = (By.CLASS_NAME, "shopping_cart_link")
    BURGER_MENU         = (By.ID, "react-burger-menu-btn")
    LOGOUT_LINK         = (By.ID, "logout_sidebar_link")

    def open(self):
        super().open("/inventory.html")

    def get_item_names(self) -> list[str]:
        return [el.text for el in self.driver.find_elements(*self.ITEM_NAMES)]

    def get_item_prices(self) -> list[float]:
        return [
            float(el.text.replace("$", ""))
            for el in self.driver.find_elements(*self.ITEM_PRICES)
        ]

    def add_item_to_cart(self, item_name: str):
        """Click 'Add to cart' for a specific product by name."""
        items = self.driver.find_elements(*self.INVENTORY_ITEMS)
        for item in items:
            name_el = item.find_element(By.CLASS_NAME, "inventory_item_name")
            if name_el.text == item_name:
                btn = item.find_element(By.XPATH, ".//button[contains(@class,'btn_inventory')]")
                btn.click()
                return
        raise ValueError(f"Item '{item_name}' not found on inventory page")

    def get_cart_count(self) -> int:
        if not self.is_visible(*self.CART_BADGE):
            return 0
        return int(self.get_text(*self.CART_BADGE))

    def sort_by(self, option: str):
        """option: 'az', 'za', 'lohi', 'hilo'"""
        Select(self.find(*self.SORT_DROPDOWN)).select_by_value(option)

    def go_to_cart(self):
        self.click(*self.CART_ICON)

    def logout(self):
        import time
        self.click(*self.BURGER_MENU)
        time.sleep(0.5)  # Wait for sidebar animation
        # JS click bypasses headless animation timing issue
        logout_el = self.find(*self.LOGOUT_LINK)
        self.driver.execute_script("arguments[0].click();", logout_el)

File: playwright-portfolio/tests/pages/inventory.page.ts
import type { Locator, Page } from '@playwright/test';
import { BasePage } from './base.page.js';

export type SortOption = 'az' | 'za' | 'lohi' | 'hilo';

export class InventoryPage extends BasePage {
  readonly path = '/inventory.html';

  readonly container: Locator;
  readonly items: Locator;
  readonly itemNames: Locator;
  readonly itemPrices: Locator;
  readonly sortDropdown: Locator;
  readonly cartBadge: Locator;
  readonly cartLink: Locator;
  readonly burgerMenu: Locator;
  readonly logoutLink: Locator;

  constructor(page: Page) {
    super(page);
    this.container = page.locator('#inventory_container');
    this.items = page.locator('.inventory_item');
    this.itemNames = page.locator('.inventory_item_name');
    this.itemPrices = page.locator('.inventory_item_price');
    this.sortDropdown = page.locator('[data-test="product-sort-container"]');
    this.cartBadge = page.locator('.shopping_cart_badge');
    this.cartLink = page.locator('.shopping_cart_link');
    this.burgerMenu = page.locator('#react-burger-menu-btn');
    this.logoutLink = page.locator('#logout_sidebar_link');
  }

  async addToCart(itemName: string): Promise<void> {
    const item = this.items.filter({ hasText: itemName });
    await item.getByRole('button', { name: /add to cart/i }).click();
  }

  async removeFromCart(itemName: string): Promise<void> {
    const item = this.items.filter({ hasText: itemName });
    await item.getByRole('button', { name: /remove/i }).click();
  }

  async cartCount(): Promise<number> {
    if ((await this.cartBadge.count()) === 0) return 0;
    const text = (await this.cartBadge.textContent()) ?? '0';
    return parseInt(text.trim(), 10);
  }

  async sortBy(option: SortOption): Promise<void> {
    await this.sortDropdown.selectOption(option);
  }

  async itemNameList(): Promise<string[]> {
    return (await this.itemNames.allTextContents()).map((s) => s.trim());
  }

  async itemPriceList(): Promise<number[]> {
    const raw = await this.itemPrices.allTextContents();
    return raw.map((s) => parseFloat(s.replace('$', '').trim()));
  }

  async openCart(): Promise<void> {
    await this.cartLink.click();
  }

  async logout(): Promise<void> {
    await this.burgerMenu.click();
    await this.logoutLink.click();
  }
}
